// flagpole/dist/logging/heading.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logitem_1 = require("./logitem");
const consoleline_1 = require("./consoleline");
const enums_1 = require("../enums");
class LogSuiteHeading extends logitem_1.LogItem {
    constructor() {
        super(...arguments);
        this.type = enums_1.LogItemType.Heading;
        this.className = "heading";
    }
    toHtml() {
        return `
            <h1>${this.message}</h1>
        `;
    }
}
exports.LogSuiteHeading = LogSuiteHeading;
class LogScenarioHeading extends LogSuiteHeading {
    toHtml() {
        return `
            <h2>${this.message}</h2>
        `;
    }
    toConsole() {
        return [new consoleline_1.SubheadingLine(this.message)];
    }
}
exports.LogScenarioHeading = LogScenarioHeading;
class LogScenarioSubHeading extends LogSuiteHeading {
    toHtml() {
        return `
            <li>
                <strong>${this.message}</strong>
            </li>
        `;
    }
    toConsole() {
        return [new consoleline_1.SectionHeadingLine(this.message)];
    }
}
exports.LogScenarioSubHeading = LogScenarioSubHeading;
