// flagpole/dist/logging/logitem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const enums_1 = require("../enums");
const consoleline_1 = require("./consoleline");
class LogItem {
    constructor(message) {
        this.timestamp = new Date();
        this.message = String(message);
    }
    get passed() {
        return false;
    }
    get failed() {
        return false;
    }
    get isOptional() {
        return false;
    }
    toString() {
        return `${this.message}`;
    }
    toConsole() {
        return [
            new consoleline_1.CustomLine(this.message, enums_1.ConsoleColor.FgWhite)
        ];
    }
    toHtml() {
        return `
            <li class="${this.className}">
                <span class="message">${this.message}</span>
            </li>
        `;
    }
    toJson() {
        return {
            timestamp: this.timestamp,
            type: this.className,
            message: this.message
        };
    }
    toCsv() {
        return `"${this.timestamp.toUTCString()}","${this.className}",` +
            `"${String(this.message).replace(/"/g, '""')}"`;
    }
    toTsv() {
        return `${this.timestamp.toUTCString()}\t${this.className}\t${this.message}`;
    }
    toPsv() {
        return `${this.timestamp.toUTCString()}|${this.className}|${this.message}`;
    }
}
exports.LogItem = LogItem;
