// flagpole/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ScenarioStatusEvent;
(function (ScenarioStatusEvent) {
    ScenarioStatusEvent[ScenarioStatusEvent["beforeExecute"] = 0] = "beforeExecute";
    ScenarioStatusEvent[ScenarioStatusEvent["executionProgress"] = 1] = "executionProgress";
    ScenarioStatusEvent[ScenarioStatusEvent["afterExecute"] = 2] = "afterExecute";
    ScenarioStatusEvent[ScenarioStatusEvent["finished"] = 3] = "finished";
})(ScenarioStatusEvent = exports.ScenarioStatusEvent || (exports.ScenarioStatusEvent = {}));
var SuiteStatusEvent;
(function (SuiteStatusEvent) {
    SuiteStatusEvent[SuiteStatusEvent["beforeAllExecute"] = 0] = "beforeAllExecute";
    SuiteStatusEvent[SuiteStatusEvent["beforeEachExecute"] = 1] = "beforeEachExecute";
    SuiteStatusEvent[SuiteStatusEvent["afterEachExecute"] = 2] = "afterEachExecute";
    SuiteStatusEvent[SuiteStatusEvent["afterAllExecute"] = 3] = "afterAllExecute";
    SuiteStatusEvent[SuiteStatusEvent["finished"] = 4] = "finished";
})(SuiteStatusEvent = exports.SuiteStatusEvent || (exports.SuiteStatusEvent = {}));
var ResponseType;
(function (ResponseType) {
    ResponseType[ResponseType["html"] = 0] = "html";
    ResponseType[ResponseType["json"] = 1] = "json";
    ResponseType[ResponseType["image"] = 2] = "image";
    ResponseType[ResponseType["stylesheet"] = 3] = "stylesheet";
    ResponseType[ResponseType["script"] = 4] = "script";
    ResponseType[ResponseType["video"] = 5] = "video";
    ResponseType[ResponseType["audio"] = 6] = "audio";
    ResponseType[ResponseType["resource"] = 7] = "resource";
    ResponseType[ResponseType["browser"] = 8] = "browser";
    ResponseType[ResponseType["extjs"] = 9] = "extjs";
})(ResponseType = exports.ResponseType || (exports.ResponseType = {}));
var LogItemType;
(function (LogItemType) {
    LogItemType[LogItemType["Result"] = 0] = "Result";
    LogItemType[LogItemType["Comment"] = 1] = "Comment";
    LogItemType[LogItemType["Heading"] = 2] = "Heading";
    LogItemType[LogItemType["SummaryData"] = 3] = "SummaryData";
})(LogItemType = exports.LogItemType || (exports.LogItemType = {}));
var ConsoleLineType;
(function (ConsoleLineType) {
    ConsoleLineType[ConsoleLineType["Pass"] = 0] = "Pass";
    ConsoleLineType[ConsoleLineType["Fail"] = 1] = "Fail";
    ConsoleLineType[ConsoleLineType["Comment"] = 2] = "Comment";
    ConsoleLineType[ConsoleLineType["Detail"] = 3] = "Detail";
    ConsoleLineType[ConsoleLineType["Heading"] = 4] = "Heading";
    ConsoleLineType[ConsoleLineType["Subheading"] = 5] = "Subheading";
    ConsoleLineType[ConsoleLineType["Decoration"] = 6] = "Decoration";
})(ConsoleLineType = exports.ConsoleLineType || (exports.ConsoleLineType = {}));
var ConsoleColor;
(function (ConsoleColor) {
    ConsoleColor["Reset"] = "\u001B[0m";
    ConsoleColor["Highlight"] = "\u001B[7m";
    ConsoleColor["Unhighlight"] = "\u001B[27m";
    ConsoleColor["FgBlack"] = "\u001B[30m";
    ConsoleColor["FgRed"] = "\u001B[31m";
    ConsoleColor["FgGreen"] = "\u001B[32m";
    ConsoleColor["FgYellow"] = "\u001B[33m";
    ConsoleColor["FgBlue"] = "\u001B[34m";
    ConsoleColor["FgMagenta"] = "\u001B[35m";
    ConsoleColor["FgCyan"] = "\u001B[36m";
    ConsoleColor["FgWhite"] = "\u001B[37m";
    ConsoleColor["FgGray"] = "\u001B[90m";
    ConsoleColor["FgBrightBlack"] = "\u001B[30;1m";
    ConsoleColor["FgBrightRed"] = "\u001B[31;1m";
    ConsoleColor["FgBrightGreen"] = "\u001B[32;1m";
    ConsoleColor["FgBrightYellow"] = "\u001B[33;1m";
    ConsoleColor["FgBrightBlue"] = "\u001B[34;1m";
    ConsoleColor["FgBrightMagenta"] = "\u001B[35;1m";
    ConsoleColor["FgBrightCyan"] = "\u001B[36;1m";
    ConsoleColor["FgBrightWhite"] = "\u001B[37;1m";
    ConsoleColor["BgBlack"] = "\u001B[40m";
    ConsoleColor["BgRed"] = "\u001B[41m";
    ConsoleColor["BgGreen"] = "\u001B[42m";
    ConsoleColor["BgYellow"] = "\u001B[43m";
    ConsoleColor["BgBlue"] = "\u001B[44m";
    ConsoleColor["BgMagenta"] = "\u001B[45m";
    ConsoleColor["BgCyan"] = "\u001B[46m";
    ConsoleColor["BgWhite"] = "\u001B[47m";
    ConsoleColor["BgBrightBlack"] = "\u001B[40;1m";
    ConsoleColor["BgBrightRed"] = "\u001B[41;1m";
    ConsoleColor["BgBrightGreen"] = "\u001B[42;1m";
    ConsoleColor["BgBrightYellow"] = "\u001B[43;1m";
    ConsoleColor["BgBrightBlue"] = "\u001B[44;1m";
    ConsoleColor["BgBrightMagenta"] = "\u001B[45;1m";
    ConsoleColor["BgBrightCyan"] = "\u001B[46;1m";
    ConsoleColor["BgBrightWhite"] = "\u001B[47;1m";
})(ConsoleColor = exports.ConsoleColor || (exports.ConsoleColor = {}));
