// flagpole/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertion_1 = require("./assertion");
exports.Assertion = assertion_1.Assertion;
const assertioncontext_1 = require("./assertioncontext");
exports.AssertionContext = assertioncontext_1.AssertionContext;
const assertionresult_1 = require("./logging/assertionresult");
exports.AssertionResult = assertionresult_1.AssertionResult;
const assertionschema_1 = require("./assertionschema");
exports.AssertionSchema = assertionschema_1.AssertionSchema;
const browsercontrol_1 = require("./browsercontrol");
exports.BrowserControl = browsercontrol_1.BrowserControl;
const browserresponse_1 = require("./browserresponse");
exports.BrowserResponse = browserresponse_1.BrowserResponse;
const cssresponse_1 = require("./cssresponse");
exports.CssResponse = cssresponse_1.CssResponse;
const domelement_1 = require("./domelement");
exports.DOMElement = domelement_1.DOMElement;
const extjsresponse_1 = require("./extjsresponse");
exports.ExtJSResponse = extjsresponse_1.ExtJSResponse;
const htmlresponse_1 = require("./htmlresponse");
exports.HtmlResponse = htmlresponse_1.HtmlResponse;
const httpresponse_1 = require("./httpresponse");
exports.HttpResponse = httpresponse_1.HttpResponse;
const imageresponse_1 = require("./imageresponse");
exports.ImageResponse = imageresponse_1.ImageResponse;
const jpath_1 = require("./jpath");
exports.jPath = jpath_1.jPath;
const jsonresponse_1 = require("./jsonresponse");
exports.JsonResponse = jsonresponse_1.JsonResponse;
const resourceresponse_1 = require("./resourceresponse");
exports.ResourceResponse = resourceresponse_1.ResourceResponse;
const response_1 = require("./response");
exports.ProtoResponse = response_1.ProtoResponse;
const scenario_1 = require("./scenario");
exports.Scenario = scenario_1.Scenario;
const scriptresponse_1 = require("./scriptresponse");
exports.ScriptResponse = scriptresponse_1.ScriptResponse;
const suite_1 = require("./suite");
exports.Suite = suite_1.Suite;
const value_1 = require("./value");
exports.Value = value_1.Value;
const videoresponse_1 = require("./videoresponse");
exports.VideoResponse = videoresponse_1.VideoResponse;
const flagpoleexecutionoptions_1 = require("./flagpoleexecutionoptions");
exports.FlagpoleExecutionOptions = flagpoleexecutionoptions_1.FlagpoleExecutionOptions;
exports.FlagpoleOutput = flagpoleexecutionoptions_1.FlagpoleOutput;
const flagpole_1 = require("./flagpole");
exports.Flagpole = flagpole_1.Flagpole;
const enums_1 = require("./enums");
exports.ResponseType = enums_1.ResponseType;
const httprequest_1 = require("./httprequest");
exports.HttpRequest = httprequest_1.HttpRequest;
const Promise = require("bluebird");
global.Promise = Promise;
const fp = flagpole_1.Flagpole;
exports.fp = fp;
const suite = (...args) => {
    return flagpole_1.Flagpole.suite(args[0], args[1]);
};
exports.suite = suite;
